File: practice-projects/gin-mvc/internal/controllers/order/order_controller.go
package order

import (
	"context"
	"strconv"

	ordersvc "gin-mvc/internal/services/order"
	"gin-mvc/pkg/response"

	"github.com/gin-gonic/gin"
)

type Controller struct {
	service *ordersvc.Service
}

func New(service *ordersvc.Service) *Controller {
	return &Controller{service: service}
}

type createOrderRequest struct {
	UserID      int64   `json:"user_id" binding:"required,gt=0"`
	TotalAmount float64 `json:"total_amount" binding:"required,gt=0"`
}

func (ctl *Controller) RegisterRoutes(api *gin.RouterGroup) {
	orders := api.Group("/orders")
	orders.POST("", ctl.CreateOrder)
	orders.GET("", ctl.GetAllOrders)
	orders.GET("/:id", ctl.GetOrder)
	orders.PUT("/:id/pay", ctl.PayOrder)
	orders.PUT("/:id/ship", ctl.ShipOrder)
	orders.PUT("/:id/deliver", ctl.DeliverOrder)
	orders.PUT("/:id/cancel", ctl.CancelOrder)
	orders.PUT("/:id/refund", ctl.RefundOrder)
}

func (ctl *Controller) CreateOrder(c *gin.Context) {
	var req createOrderRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, err.Error())
		return
	}
	o, err := ctl.service.CreateOrder(c.Request.Context(), req.UserID, req.TotalAmount)
	if err != nil {
		response.Error(c, 2001, err.Error())
		return
	}
	response.SuccessWithMessage(c, "订单创建成功", o)
}

func (ctl *Controller) GetOrder(c *gin.Context) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		response.BadRequest(c, "invalid order id")
		return
	}
	o, err := ctl.service.GetOrderByID(c.Request.Context(), id)
	if err != nil {
		response.Error(c, 2001, err.Error())
		return
	}
	response.Success(c, o)
}

func (ctl *Controller) GetAllOrders(c *gin.Context) {
	orders, err := ctl.service.GetAllOrders(c.Request.Context())
	if err != nil {
		response.InternalServerError(c, err.Error())
		return
	}
	response.Success(c, orders)
}

func (ctl *Controller) GetUserOrders(c *gin.Context) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		response.BadRequest(c, "invalid user id")
		return
	}
	orders, err := ctl.service.GetUserOrders(c.Request.Context(), id)
	if err != nil {
		response.InternalServerError(c, err.Error())
		return
	}
	response.Success(c, orders)
}

func (ctl *Controller) PayOrder(c *gin.Context) {
	ctl.updateStatusAction(c, ctl.service.PayOrder, "订单支付成功")
}

func (ctl *Controller) ShipOrder(c *gin.Context) {
	ctl.updateStatusAction(c, ctl.service.ShipOrder, "订单发货成功")
}

func (ctl *Controller) DeliverOrder(c *gin.Context) {
	ctl.updateStatusAction(c, ctl.service.DeliverOrder, "订单送达确认成功")
}

func (ctl *Controller) CancelOrder(c *gin.Context) {
	ctl.updateStatusAction(c, ctl.service.CancelOrder, "订单取消成功")
}

func (ctl *Controller) RefundOrder(c *gin.Context) {
	ctl.updateStatusAction(c, ctl.service.RefundOrder, "订单退款成功")
}

func (ctl *Controller) updateStatusAction(c *gin.Context, fn func(ctx context.Context, id int64) error, successMsg string) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		response.BadRequest(c, "invalid order id")
		return
	}
	if err := fn(c.Request.Context(), id); err != nil {
		response.Error(c, 2002, err.Error())
		return
	}
	response.SuccessWithMessage(c, successMsg, nil)
}

File: practice-projects/nestjs-ddd/src/interfaces/http/order/order.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  Post,
  Put,
} from '@nestjs/common';
import { ApiOperation, ApiTags } from '@nestjs/swagger';
import { CreateOrderCommand } from '../../../application/order/dto/order.commands';
import { OrderApplicationService } from '../../../application/order/order-application.service';
import { CreateOrderRequest } from './vo/order.request';
import { OrderResponse } from './vo/order.response';

/**
 * 订单 HTTP 接口，路由与 gin-ddd/internal/interfaces/router/order_router.go 对齐：
 *   POST   /api/orders
 *   GET    /api/orders
 *   GET    /api/orders/:id
 *   PUT    /api/orders/:id/pay
 *   PUT    /api/orders/:id/ship
 *   PUT    /api/orders/:id/deliver
 *   PUT    /api/orders/:id/cancel
 *   PUT    /api/orders/:id/refund
 */
@ApiTags('订单管理')
@Controller('orders')
export class OrderController {
  constructor(private readonly orderAppService: OrderApplicationService) {}

  @Post()
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: '创建订单' })
  async create(@Body() req: CreateOrderRequest): Promise<OrderResponse> {
    const dto = await this.orderAppService.createOrder(
      new CreateOrderCommand(req.user_id, req.total_amount),
    );
    return OrderResponse.fromDto(dto);
  }

  @Get()
  @ApiOperation({ summary: '获取所有订单' })
  async findAll(): Promise<OrderResponse[]> {
    const dtos = await this.orderAppService.getAllOrders();
    return OrderResponse.fromDtos(dtos);
  }

  @Get(':id')
  @ApiOperation({ summary: '根据 ID 获取订单' })
  async findOne(@Param('id', ParseIntPipe) id: number): Promise<OrderResponse> {
    const dto = await this.orderAppService.getOrderById(id);
    return OrderResponse.fromDto(dto);
  }

  @Put(':id/pay')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: '支付订单' })
  async pay(@Param('id', ParseIntPipe) id: number): Promise<OrderResponse> {
    const dto = await this.orderAppService.payOrder(id);
    return OrderResponse.fromDto(dto);
  }

  @Put(':id/ship')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: '订单发货' })
  async ship(@Param('id', ParseIntPipe) id: number): Promise<OrderResponse> {
    const dto = await this.orderAppService.shipOrder(id);
    return OrderResponse.fromDto(dto);
  }

  @Put(':id/deliver')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: '确认送达' })
  async deliver(@Param('id', ParseIntPipe) id: number): Promise<OrderResponse> {
    const dto = await this.orderAppService.deliverOrder(id);
    return OrderResponse.fromDto(dto);
  }

  @Put(':id/cancel')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: '取消订单' })
  async cancel(@Param('id', ParseIntPipe) id: number): Promise<OrderResponse> {
    const dto = await this.orderAppService.cancelOrder(id);
    return OrderResponse.fromDto(dto);
  }

  @Put(':id/refund')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: '订单退款' })
  async refund(@Param('id', ParseIntPipe) id: number): Promise<OrderResponse> {
    const dto = await this.orderAppService.refundOrder(id);
    return OrderResponse.fromDto(dto);
  }
}
